### DISCUSSION TITLE
(React Router) Outlet context for child route has previous data upon navigation

### DISCUSSION THREAD

**Discussion by ckpearson** [remix-run/remix#6012] [Q&A]
I'm (relatively) new to react, and have spent some time bashing my head around nested routing, and I thought I had a grasp on it, but I'm seeing weird behaviour with `useOutletContext()` that I can only explain by reasoning that the child component isn't being re-mounted on navigation.

For the purposes of this, let's assume the following minimal repro:

### Routes
```tsx
<Route path="/" Component={App}>
  <Route path="outers">
    <Route path=":outerId" element={<OuterShell />}>
      <Route index element={<InnerIndex />} />
      </Route>
    </Route>
  </Route>
</Route>
```

### OuterShell
```tsx
const OuterShell = () => {
  const params = useParams();

  const [ctx, setCtx] = useState({
    message: `Hello for ${JSON.stringify(params)}`,
  });

  return (
    <div>
      <p>Shell</p>
      <p>{JSON.stringify(params)}</p>
      <Outlet context={ctx} />
    </div>
  );
};
```

### InnerIndex
```tsx
const InnerIndex = () => {
  const params = useParams();
  const ctx = useOutletContext<{ message: string }>();
  return (
    <div>
      <p>Index</p>
      <p>{JSON.stringify(params)}</p>
      <p>{JSON.stringify(ctx)}</p>
    </div>
  );
};
```

My intuition is that when navigating to anything at `/outers/{id}` I would be presented with:
```
Shell
{ outerId: 'the-id' }
Index
{ outerId: 'the-id' }
{ message: 'Hello for { outerId: 'the-id' } ' }
```
And in fact that's what I see when HMR forces a refresh, or I manually refresh the browser, but if I happen to be on one of the InnerIndex routes, if I then navigate to another say from `outers/a` to `outers/b`, what I get instead is:

```
Shell
{ outerId: 'b' }
Index
{ outerId: 'b' }
{ message: 'Hello for { outerId: 'a } ' }
```

I'm confused though, as the second navigation's `getParams` call is clearly being invoked, so I don't understand why `useOutletContext` call wouldn't be retrieving the new context value from `OuterShell` on the `outers/b` navigation.

Like I said, my first intuition was that the `InnerIndex` component wasn't being re-mounted / re-rendered on the second call, but come to think of it that doesn't make much sense given that the `getParams()` call is clearly working.

**Comment by ckpearson** [comment]
I've discovered I can "fix" the issue by doing the following:

### OuterShell
```tsx
const OuterShell = () => {
  const params = useParams();

  const constructCtxState = () => {
    return {
      message: `Hello for ${JSON.stringify(params)}`,
    };
  };

  const [ctx, setCtx] = useState(constructCtxState());

  useEffect(() => {
    setCtx(constructCtxState());
  }, [params, setCtx]);

  return (
    <div>
      <p>Shell</p>
      <p>{JSON.stringify(params)}</p>
      <Outlet context={ctx} />
    </div>
  );
};
```

By using `useEffect` to force setting a new context value when `params` changes, this does now produce the desired behaviour, but I can't help but feel like this is the wrong way to go about it.

**Reply by ckpearson** [reply]
Okay, so I've since slept, and had a think about it, and I think this is less of an issue with react router and just the reality of state management within React.

I *believe* that what's happening is that the `InnerIndex` component is staying mounted on subsequent navigations, and is being re-rendered, which means the params are updated, but because the `useState` hook call knows that there is already a value for this instance of the component, it just ignores the new correct initial value, and echoes out the old state value.

So, actually I think the `useEffect` approach to draw a direct relationship between the context value and the params values is the correct way of doing it, and it just feels unintuitive.

**Reply by kiliman** [reply]
What you really want to do is set a `key` prop on your `<Outlet>` and set it to the value that changes based on the route. It could be as simple as `useLocation().pathname`

This will force React to unmount the `<Outlet>` on route changes.
